### ISSUE TITLE
Job.next_run_time not available form the outside

### ISSUE THREAD

**Issue by Hendiadyoin1** [agronholm/apscheduler#434]
I'm working with the scheduler in a Django App (should not affect behavior)

I create a Job to Backup at specific Intervals and want to log the next execution time.
Now i log it with:
```python
logger.info(f"next Backup at {scheduler.get_job(job_id).next_run_time.strftime('%Y-%m-%d %H:%M:%S')}") 
```
this raises an attribute Error:

>     logger.info(f"next Backup at {scheduler.get_job(job_id).next_run_time.strftime('%Y-%m-%d %H:%M:%S')}") 
> AttributeError: next_run_time

but when just logging the job i get a right output:
> Backup.backup (trigger: interval[0:10:00], next run at: 2020-05-15 17:49:26 CEST)

when looking in the code i see that the job uses `self.next_run_time`

so the attribute is surely available, but not accessible as it seems
in addition to that is the AttributeError incomplete just saying the attribute, not that this attribute does not exist, which seems pretty wierd...

complete relevant code below code below:
```python
 def startSubroutine(self)->None:
        """
        Start the backup-subroutine
        """
        self.start_time = datetime.datetime.now()
        self.scheduler = BackgroundScheduler()
        interval = self.settings['interval']
        self.job_id = self.scheduler.add_job(self.backup, 'interval', **interval).id

        # prepare info string to log
        info_str = format_interval("Scheduled backup for ", interval)

        logger.info(info_str)
        # the critical line:
        logger.info(f"Next Backup at {self.scheduler.get_job(self.job_id).next_run_time.strftime('%Y-%m-%d %H:%M:%S')}")

        # start subroutine
        self.scheduler.start()
```

**Comment by AlexandreLI** [comment]
This one work for me:
self.job_id = self.scheduler.add_job(self.backup, 'interval', **interval, id='daily_backup')
logging.info("next Backup at %s" % self.scheduler.get_job("daily_backup").next_run_time)

**Comment by Hendiadyoin1** [comment]
So further looking into it i see that the next_run_time isn't always set, but gets set at some point by the Scheduler or something ub the __setstate__ method
Then is it possible to trigger the calculation of the job. because just getting the job does not seem to update it

**Comment by AnjeKnottnerus** [comment]
One way hacky way I accessed next_run_time, is by adding a listener with a function that is updating a dictionary every time a EVENT_JOB_SUBMITTED is noticed.

This is the listerer function: 
`from apscheduler.schedulers.background import BlockingScheduler
from apscheduler.events import EVENT_JOB_SUBMITTED

# Place to store scheduled_time of each run
current_run = {"scheduled_time": None}


def _update_scheduled_time(event):
    """Updates the scheduledtime in config based on
    a JobSubmissionEvent.

    Args:
        event (apscheduler.events): JobSubmissionEvent
    """
    current_run["scheduled_time"] = event.scheduled_run_times`

Here I add the listener and the job to the ap.Scheduler object: 

`    # ----------------------------------------
    # Create scheduler job and listener
    # ----------------------------------------
    # Create scheduler object
    sched = BlockingScheduler()

    # Add a listerer that listents to EVENT_JOB_SUBMITTED and updates scheduled_time
    sched.add_listener(_update_scheduled_time, EVENT_JOB_SUBMITTED)

    # Add schedulerjob
    sched.add_job(
        scheduled_job.produce_request, **scheduled_job.get_job_specification(), kwargs={"current_run": current_run}
    )`

The current_run dict passed with kwargs is updated every time a new job is submitted. Since function passed to the listener can only take one argument (the event) and cannot return a value, I am updating a dictionary in global context.

**Comment by MatthewShotton** [comment]
I know this is an old issue but dropping my solution here in case it helps anyone who googles this. I faced a similar problem and have found a solution which seems to work quite nicely.

It looks like `next_run_time` is not available until the job is scheduled (you can check the `pending` property on the job instance to see if it's scheduled or not). The speed at which the job get's scheduled depends on the scheduler you use.

An alternative way to get the `next_run_time` is to call `add_job` with an explicit `Trigger` object. Then you can call `get_next_fire_time()` on the `Trigger` object at any time to get the time the `Trigger` will next cause the job to run.

A minimal example:

```Python
import datetime
from apscheduler.schedulers.blocking import BlockingScheduler
from apscheduler.triggers.cron import CronTrigger

def job_func():
	print("Doing a job")

scheduler = BlockingScheduler()

# Trigger every even minute
trigger = CronTrigger(hour="*", minute="*/2")

# Add the job
scheduler.add_job(job_func, trigger)

# Get the next time the job will trigger, you can call this at anytime.
next_fire_time = trigger.get_next_fire_time(None, datetime.datetime.now())

print("Job next running @ ", next_fire_time)

# Start the scheduler so the job will get scheduled.
scheduler.start()
```

**Comment by agronholm** [comment]
The job's scheduled run time is available on v4.0 via `apscheduler.current_job`:
```python
from apscheduler import current_job

def my_task():
    print("This job was scheduled to run at", current_job.get().scheduled_fire_time)
```
Note that the API may change before v4.0 final.
